src/query/service/src/sql/optimizer/cascades/mod.rs

```diff
@@ -15,115 +15,119 @@
 mod explore_rules;
 mod implement_rules;
 
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
 use common_exception::ErrorCode;
 use common_exception::Result;
 
+use super::cost::Cost;
+use super::cost::CostContext;
+use super::cost::CostModel;
+use super::cost::DefaultCostModel;
 use crate::sql::optimizer::cascades::explore_rules::get_explore_rule_set;
 use crate::sql::optimizer::cascades::implement_rules::get_implement_rule_set;
-use crate::sql::optimizer::group::Group;
+use crate::sql::optimizer::format::display_memo;
 use crate::sql::optimizer::m_expr::MExpr;
 use crate::sql::optimizer::memo::Memo;
-use crate::sql::optimizer::rule::RulePtr;
 use crate::sql::optimizer::rule::RuleSet;
 use crate::sql::optimizer::rule::TransformState;
-use crate::sql::optimizer::RequiredProperty;
 use crate::sql::optimizer::SExpr;
 use crate::sql::plans::Operator;
 use crate::sql::IndexType;
 
 /// A cascades-style search engine to enumerate possible alternations of a relational expression and
 /// find the optimal one.
-///
-/// NOTICE: we don't support cost-based optimization and lower bound searching for now.
-#[allow(dead_code)]
 pub struct CascadesOptimizer {
-    // optimize_context: OptimizeContext,
     memo: Memo,
     explore_rules: RuleSet,
     implement_rules: RuleSet,
+
+    cost_model: Box<dyn CostModel>,
+
+    /// group index -> best cost context
+    best_cost_map: HashMap<IndexType, CostContext>,
 }
 
 impl CascadesOptimizer {
-    #[allow(dead_code)]
     pub fn create() -> Self {
         CascadesOptimizer {
             memo: Memo::create(),
             explore_rules: get_explore_rule_set(),
             implement_rules: get_implement_rule_set(),
+            cost_model: Box::new(DefaultCostModel),
+            best_cost_map: HashMap::new(),
         }
     }
 
-    #[allow(dead_code)]
     fn init(&mut self, expression: SExpr) -> Result<()> {
```

Comment:
I noticed you changed `expression` to `s_expr` in other places, maybe you'll be pleased to change it here

Reply:
Fixed.